src/string_id.h

```diff
@@ -69,69 +108,92 @@ template<typename T>
+        template<unsigned N>
+        explicit constexpr string_id( const char ( &str )[N] ) : _hash( Hash::CompileTime::fnv1_64( str ) ),
+            _char_ptr( str ), _is_char_ptr( true ) {}
+
+        template<typename U>
+        string_id( const string_id<U> &other ) : _hash( other.hash() ), _char_ptr( other._char_ptr ),
+            _is_char_ptr( other._is_char_ptr ) {}
+
         template<typename S, class = typename
-                 std::enable_if< std::is_convertible<S, std::string >::value>::type >
-        explicit string_id( S &&
-                            id ) : _id( std::forward<S>( id ) ), _cid( INVALID_CID ), _version( INVALID_VERSION ) {}
+                 std::enable_if< std::is_convertible< S, std::string >::value >::type
+                 >
+        explicit string_id( S && id ) {
+            auto str = std::make_unique<std::string>( std::forward<S>( id ) );
+            uint64_t hash = Hash::RunTime::fnv1_64( str->c_str(), str->size() );
+            _hash = hash;
+
+            auto insertion = cached_strings().try_emplace( hash, std::move( str ) );
+            auto cached_string = insertion.first->second.get();
+
+            // Verify we don't have hash collisions
+            //assert(insertion.second || *cached_string == id);
+
+            _string = cached_string;
+        }
+
         /**
          * Default constructor constructs an empty id string.
          * Note that this id class does not enforce empty id strings (or any specific string at all)
          * to be special. Every string (including the empty one) may be a valid id.
          */
-        string_id() : _cid( INVALID_CID ), _version( INVALID_VERSION ) {}
+        constexpr string_id() : string_id( "" ) {};
         /**
          * Comparison, only useful when the id is used in std::map or std::set as key. Compares
          * the string id as with the strings comparison.
          */
         bool operator<( const This &rhs ) const {
-            return _id < rhs._id;
+            // Not str() compares to prevent needless tier ups to std::string
+            return strcmp( c_str(), rhs.c_str() ) < 0;
         }
         /**
          * The usual comparator, compares the string id as usual.
          */
         bool operator==( const This &rhs ) const {
-            bool can_compare_cid = ( _cid != INVALID_CID || rhs._cid != INVALID_CID ) &&
-                                   _version == rhs._version && _version != INVALID_VERSION;
-            return ( can_compare_cid && _cid == rhs._cid ) ||
-                   ( !can_compare_cid && _id == rhs._id );
-            // else returns false, when:
-            //      can_compare_cid && _cid != rhs._cid
-            //  OR  !can_compare_cid && _id != rhs._id
+            return _hash == rhs._hash;
+        }
+        /**
+         * Allows to quickly compare hashes, when caching the hash would be faster
+         */
+        bool operator==( uint64_t hash ) const {
+            return _hash == hash;
```

Comment:
If we are to completely avoid hash collision then this function should be removed.